File: codeForces_new/F_Wildflower.py
import sys
sys.setrecursionlimit(10**7)
input = sys.stdin.readline

MOD = 10**9 + 7

def modexp(a, e=1, mod=MOD):
    res = 1
    while e:
        if e & 1:
            res = (res * a) % mod
        a = (a * a) % mod
        e >>= 1
    return res

def solve():
    t = int(input())
    for _ in range(t):
        n = int(input())
        g = [[] for _ in range(n+1)]
        for _ in range(n-1):
            u,v = map(int, input().split())
            g[u].append(v)
            g[v].append(u)

        # Build rooted tree at 1
        parent = [0]*(n+1)
        children = [[] for _ in range(n+1)]
        parent[1] = -1
        stack = [1]
        order = []
        while stack:
            u = stack.pop()
            order.append(u)
            for w in g[u]:
                if parent[w]==0:
                    parent[w]=u
                    children[u].append(w)
                    stack.append(w)

        # Compute subtree sizes bottom-up
        size = [1]*(n+1)
        for u in reversed(order):
            for w in children[u]:
                size[u] += size[w]

        ok = True
        # Check each node
        for u in range(1, n+1):
            c = children[u]
            if len(c) >= 3:
                ok = False
                break
            if len(c) == 2:
                # must split equally
                if size[c[0]] != size[c[1]]:
                    ok = False
                    break
            # len(c)==0 or 1 always fine

        if not ok:
            print(0)
        else:
            # Every labeling works
            print(modexp(2, n))

if __name__ == "__main__":
    solve()

File: codeForces_new/F_Wildflower.java
import java.io.*;
import java.util.*;

public class F_Wildflower {
    static final int MOD = 1_000_000_007;

    // Fast exponentiation modulo MOD
    static long modexp(long a, long e) {
        long res = 1;
        while (e > 0) {
            if ((e & 1) == 1) {
                res = (res * a) % MOD;
            }
            a = (a * a) % MOD;
            e >>= 1;
        }
        return res;
    }

    public static void main(String[] args) throws IOException {
        // BufferedReader for fast input
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();

        while (t-- > 0) {
            int n = Integer.parseInt(br.readLine());
            ArrayList<Integer>[] g = new ArrayList[n + 1];
            for (int i = 1; i <= n; i++) {
                g[i] = new ArrayList<>();
            }

            for (int i = 0; i < n - 1; i++) {
                String[] parts = br.readLine().split(" ");
                int u = Integer.parseInt(parts[0]);
                int v = Integer.parseInt(parts[1]);
                g[u].add(v);
                g[v].add(u);
            }

            // Build rooted tree at 1
            int[] parent = new int[n + 1];
            ArrayList<Integer>[] children = new ArrayList[n + 1];
            for (int i = 1; i <= n; i++) {
                children[i] = new ArrayList<>();
            }
            parent[1] = -1;

            Stack<Integer> stack = new Stack<>();
            stack.push(1);
            ArrayList<Integer> order = new ArrayList<>();

            while (!stack.isEmpty()) {
                int u = stack.pop();
                order.add(u);
                for (int w : g[u]) {
                    if (parent[w] == 0) {
                        parent[w] = u;
                        children[u].add(w);
                        stack.push(w);
                    }
                }
            }

            // Compute subtree sizes bottom-up
            int[] size = new int[n + 1];
            Arrays.fill(size, 1);
            for (int i = order.size() - 1; i >= 0; i--) {
                int u = order.get(i);
                for (int w : children[u]) {
                    size[u] += size[w];
                }
            }

            boolean ok = true;
            for (int u = 1; u <= n; u++) {
                int csize = children[u].size();
                if (csize >= 3) {
                    ok = false;
                    break;
                }
                if (csize == 2) {
                    int c0 = children[u].get(0);
                    int c1 = children[u].get(1);
                    if (size[c0] != size[c1]) {
                        ok = false;
                        break;
                    }
                }
            }

            if (!ok) {
                sb.append("0\n");
            } else {
                sb.append(modexp(2, n)).append("\n");
            }
        }

        System.out.print(sb);
    }
}
